pkg/operator/metricscontroller/legacy_cn_certs_sync.go

```diff
@@ -20,42 +21,54 @@ import (
 // The supported prometheus query result types are:
 // - vector: only the first sample of the vector will be used for evaluation.
 // - scalar: the returned value will be used.
-func NewLegacyCNCertsMetricsSyncFunc(query string, operatorClient v1helpers.OperatorClient) MetricsSyncFunc {
+func NewLegacyCNCertsMetricsSyncFunc(conditionPrefix, query string, operatorClient v1helpers.OperatorClient) MetricsSyncFunc {
 	return func(ctx context.Context, controllerContext factory.SyncContext, client prometheusv1.API) error {
 		result, _, err := client.Query(ctx, query, time.Now())
 		if err != nil {
 			return fmt.Errorf("error executing prometheus query: %w", err)
 		}
 
-		var value model.SampleValue
+		var samples model.Vector
+
 		switch result.Type() {
 		case model.ValVector:
-			vec := result.(model.Vector)
-			if len(vec) == 0 {
+			samples = result.(model.Vector)
+			if len(samples) == 0 {
 				return fmt.Errorf("empty vector result from query: %q", query)
 			}
-			value = vec[0].Value
 		case model.ValScalar:
-			value = result.(*model.Scalar).Value
+			val := result.(*model.Scalar)
+			samples = append(samples, &model.Sample{
+				Metric:    nil,
+				Value:     val.Value,
+				Timestamp: val.Timestamp,
+			})
 		default:
 			return fmt.Errorf("unexpected prometheus query return type: %T", result)
 		}
 
-		_, _, err = v1helpers.UpdateStatus(operatorClient, v1helpers.UpdateConditionFn(newInvalidCertsCondition(float64(value))))
+		_, _, err = v1helpers.UpdateStatus(operatorClient, v1helpers.UpdateConditionFn(newInvalidCertsCondition(conditionPrefix, samples)))
 		return err
 	}
 }
 
-func newInvalidCertsCondition(invalidCertsCount float64) operatorv1.OperatorCondition {
+func newInvalidCertsCondition(conditionPrefix string, samples model.Vector) operatorv1.OperatorCondition {
 	condition := operatorv1.OperatorCondition{
-		Type:   "InvalidCertsUpgradeable",
+		Type:   conditionPrefix + "InvalidCertsUpgradeable",
 		Status: operatorv1.ConditionTrue,
 	}
 
-	if invalidCertsCount >= 1.0 {
+	var invalidSANs []string
+	for _, sample := range samples {
+		if sample.Value > 0 {
+			invalidSANs = append(invalidSANs, fmt.Sprintf("%s => %s", sample.Metric, sample.Value))
```

Comment:
`sample.Metric` is `nil` if the orig result was of `ValScalar`. Is there a better way to init that struct based on our knowledge? Also, isn't `sample.Metric` a map? What will such output look like? Should it be `"%v => %s"` maybe?

Reply:
`%s` will gracefully print `{}` in this case, see the unit test at: https://github.com/openshift/library-go/blob/7127a1be248594d05807f87be265309c7d6b0315/pkg/operator/metricscontroller/legacy_cn_certs_sync_test.go#L136-L147

Here concretely it will print `server certificates without SAN detected: {} => 10` which is how prometheus outputs metrics without labels which applies to scalars.

Reply:
when will that happen and how am I as an administrator supposed to know what it means?

Reply:
wouldn't it make more sense for the message to read something like "server certificates without SAN were detected in webhooks and aggregated APIs"?

Reply:
This is a synthetic test case for scalars. In your concrete case of webhooks and/or aggregated APIs the message will be:
```
server certificates without SAN detected: {type="webhooks"} => 123, {type="aggregated"} => 456
```

Reply:
in the case of oauth it will be something like:
```
server certificates without SAN detected: {provider="OpenID"} => 123, {provider="LDAP"} => 456
```

Reply:
ok, my proposal is therefore:
1. do not support scalars
2. don't print the `%s => %s` in case of scalars

Also, shouldn't the message be: `times server certificates without SAN were detected: ` instead of what it currently is? IIRC the metrics increases when a server certificate was spotted not for each unique server certificate.

Reply:
> how am I as an administrator supposed to know what it means?

This question rather refers to the semantics of a concrete playbook. We can indicate the resolution in the text message here of course. How about something like:

```
Server certificates without SAN detected: {provider="OpenID"} => 123, {provider="LDAP"} => 456. These have to be replaced to include SAN fields and not rely on the CN field to be valid.
```

Reply:
plus, instead of the arrow we can make it an equals so it becomes more obvious:
```
Server certificates without SAN detected: {provider="OpenID"} = 123, {provider="LDAP"} = 456. These have to be replaced to include SAN fields and not rely on the CN field to be valid.
```

Reply:
Agreed, let's remove the support for scalars as well.